test/demo/expiry.inttest.ts

```diff
@@ -0,0 +1,296 @@
+
+import {NotificationClosedEvent, NotificationOptions, NotificationActionEvent, Notification} from '../../src/client';
+import {ActionTrigger} from '../../src/client/actions';
+import {Events} from '../../src/client/internal';
+import {createAppInServiceRealm} from '../utils/int/spawnRemote';
+import {testManagerIdentity, testAppUrlDefault} from '../utils/int/constants';
+import * as notifsRemote from '../utils/int/notificationsRemote';
+import * as providerRemote from '../utils/int/providerRemote';
+import {delay, Duration} from '../utils/int/delay';
+import {waitForAppToBeRunning} from '../utils/int/common';
+
+let testApp: Application;
+let testWindow: _Window;
+
+let eventLog: {event: Events, time: number}[];
+let closedListener: jest.Mock<void, [NotificationClosedEvent]>;
+let actionListener: jest.Mock<void, [NotificationActionEvent]>;
+
+const options: NotificationOptions = {
+    body: 'Test Notification Body',
+    title: 'Test Notification Title',
+    category: 'Test Notification Category'
+};
+
+beforeEach(async () => {
+    jest.resetAllMocks();
+
+    testApp = await createAppInServiceRealm(testManagerIdentity, {url: testAppUrlDefault});
+    testWindow = await testApp.getWindow();
+
+    eventLog = [];
+    const eventLogger = (event: Events) => eventLog.push({event, time: Date.now()});
+
+    actionListener = jest.fn<void, [NotificationActionEvent]>().mockImplementation(eventLogger);
+    closedListener = jest.fn<void, [NotificationClosedEvent]>().mockImplementation(eventLogger);
+
+    await notifsRemote.addEventListener(testWindow.identity, 'notification-closed', closedListener);
+    await notifsRemote.addEventListener(testWindow.identity, 'notification-action', actionListener);
+});
+
+afterEach(async () => {
+    await providerRemote.clearStoredNotifications(testWindow.identity);
+    await testApp.quit();
+});
+
+test('When a notification is created with an expiry, the notification is removed when the expiry is reached', async () => {
+    const expires = future(seconds(5));
+
+    await notifsRemote.createAndAwait(testWindow.identity, {...options, expires});
+
+    await delay(seconds(5));
+    await delay(Duration.EVENT_PROPAGATED);
+
+    await expect(notifsRemote.getAll(testWindow.identity)).resolves.toEqual([]);
+
+    expect(closedListener).toBeCalledTimes(1);
+    expectTimeSoonAfter(eventLog[0].time, expires.getTime());
+});
+
+test('When a notification is created with an expiry in the past, the notification is removed immediately', async () => {
+    await notifsRemote.createAndAwait(testWindow.identity, {...options, expires: past(seconds(5))});
+
+    await delay(Duration.EVENT_PROPAGATED);
+
+    await expect(notifsRemote.getAll(testWindow.identity)).resolves.toEqual([]);
+    expect(closedListener).toBeCalledTimes(1);
+});
+
+test('When a notification is created with an expiry and an expiry action, an action event is received when it expires', async () => {
```

Comment:
Do we need a "null" version of this test? No action is raised when there isn't an `onExpire`?

Could just be an extra assert in the previous test.

Reply:
Extra assert in the first two tests (similar to how we're checking the result of getAll in those two tests)

✅